src/corefx/System.Globalization.Native/locale.cpp

```diff
@@ -0,0 +1,373 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+#include <stdint.h>
+#include <string.h>
+
+#include "localeInformationConstants.hpp"
+
+#include "unicode/dcfmtsym.h" //decimal
+#include "unicode/dtfmtsym.h" //date
+#include "unicode/ulocdata.h"
+
+/*
+Function:
+GetLocale
+
+Returns a locale given the locale name
+*/
+Locale GetLocale(const UChar* localeName, bool canonize = false)
+{
+	char localeNameTemp[ULOC_FULLNAME_CAPACITY];
+
+	if (localeName != NULL)
+	{
+		int32_t len = u_strlen(localeName);
+		u_UCharsToChars(localeName, localeNameTemp, len);
+		localeNameTemp[len] = (char)0;
+	}
+
+	Locale loc;
+	if (canonize)
+	{
+		loc = Locale::createCanonical(localeName == NULL ? NULL : localeNameTemp);
+	}
+	else
+	{
+		loc = Locale::createFromName(localeName == NULL ? NULL : localeNameTemp);
+	}
+
+	return loc;
+}
+
+/*
+Function:
+GopyCharToUChar
+
+Copies the given char to UChar with error checking
+*/
+UErrorCode CopyCharToUChar(const char *str, UChar* value, int32_t valueLength)
+{
+	int len = strlen(str);
+	if (len >= valueLength)
+	{
+		return U_BUFFER_OVERFLOW_ERROR;
+	}
+	u_charsToUChars(str, value, len + 1);
+	return U_ZERO_ERROR;
+}
+
+/*
+PAL Function:
+getCanonicalLocaleName
+
+Obtains a canonical locale given the locale name
+*/
+extern "C" int32_t GetLocaleName(const UChar* localeName, UChar* value, int32_t valueLength)
+{
+	Locale locale = GetLocale(localeName, true);
```

Comment:
The C++ APIs seemed more intuitive. LMK if you want consistency here and I could switch to C. AFAIK there is nothing that can only be done on either side. Other than that, no preference although perhaps error cases would be easier in C++ because of their smart pointer (LocalPointer<T>) although currently the code doesn't use that (I plan on fixing that)